test/hummingbot/connector/exchange/ascend_ex/test_ascend_ex_api_order_book_data_source.py

```diff
@@ -0,0 +1,582 @@
+        ws_connect_mock.side_effect = Exception()
+
+        queue = asyncio.Queue()
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_order_book_diffs(ev_loop=self.ev_loop, output=queue))
+        await asyncio.sleep(0.5)
+
+        await self.data_source._shared_client.close()
+
+    def test_listen_for_order_book_diff_ws_connection_exception_details_are_logged(self):
+        self.ev_loop.run_until_complete(self.do_listen_for_order_book_diff_ws_connection_exception_details_are_logged())
+        self.assertTrue(self._is_logged("ERROR", "Unexpected error with WebSocket connection. Retrying after 30 seconds..."))
+
+    @patch('aiohttp.ClientSession.ws_connect', new_callable=AsyncMock)
+    async def do_listen_for_order_book_diff_logs_exceptions_details(self, ws_connect_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        sync_queue = asyncio.Queue()
+
+        self.data_source._trading_pairs = ["BTC-USDT"]
+        websocket_mock = self.mocking_assistant.create_websocket_mock()
+        websocket_mock.receive.side_effect = Exception()
+        websocket_mock.close.side_effect = lambda: sync_queue.put_nowait(1)
+        ws_connect_mock.return_value = websocket_mock
+
+        queue = asyncio.Queue()
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_order_book_diffs(ev_loop=self.ev_loop, output=queue))
+
+        try:
+            await asyncio.wait_for(self.mocking_assistant._all_ws_delivered.wait(), 0)
+        except asyncio.exceptions.TimeoutError:
+            # The exception will happen when cancelling the task
+            pass
+
+        await self.data_source._shared_client.close()
+
+    def test_listen_for_order_book_diff_logs_exceptions_details(self):
+        self.ev_loop.run_until_complete(self.do_listen_for_order_book_diff_logs_exceptions_details())
+        self.assertTrue(self._is_logged("ERROR", "Unexpected error with WebSocket connection. Retrying after 30 seconds..."))
+
+    @patch('aiohttp.ClientSession.ws_connect', new_callable=AsyncMock)
+    async def do_listen_for_trades(self, mock_response, ws_connect_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        self.data_source._trading_pairs = ["BTC-USDT"]
+        ws_connect_mock.return_value = self.mocking_assistant.create_websocket_mock()
+
+        trades_queue = asyncio.Queue()
+
+        task = self.ev_loop.create_task(self.data_source.listen_for_trades(ev_loop=self.ev_loop, output=trades_queue))
+
+        self.mocking_assistant.add_websocket_aiohttp_message(ws_connect_mock.return_value, json.dumps(mock_response))
+
+        # Lock the test to let the async task run
+        try:
+            await asyncio.wait_for(self.mocking_assistant._all_ws_delivered.wait(), 0)
+        except asyncio.exceptions.TimeoutError:
+            # The exception will happen when cancelling the task
+            pass
```

Comment:
We should make the test wait for the trades to be included in the queue (as done afterwards). This code seems to not be required.